Offload remote inbox notifications engine run using action-scheduler

becdetat opened the pull request:
Partially fixes #6189

This uses action-scheduler to offload running the remote inbox notifications engine run when adding or importing products.

### Detailed test instructions:

1. delete all products
5. delete all admin notes (using the test helper plugin)
2. add test feed url to `DataSourcePoller`: `https://gist.githubusercontent.com/becdetat/2d9895ad8e159ece76e8b3d029b09e8b/raw/30ad24cef0945220fbe351fa5af5d5cca8f4cf55/three-products.json`
6. add three new products
7. wait a minute for action-scheduler to do it's magic
8. visit WC Admin's home page
9. The new test note should be present

![image](https://user-images.githubusercontent.com/224531/118444713-d4b6e000-b730-11eb-9aaa-0a1f688f588e.png)

becdetat committed:
Offload remote inbox notifications engine run using action-scheduler

becdetat committed:
release note

becdetat committed:
Merge branch 'main' into dev/offload-rin-engine-run-on-product-add

becdetat committed:
fix changelog

ilyasfoo reviewed (commented):
When I tested this, I hit a couple of snags. Not all of them are related to the PR's changes but they do interfere with testing:
- The note did not come out, and I think it was because 
`wc_remote_inbox_notifications_specs` option already exists and had value from the original datasource url, thus the poller did not fetch new. I had to delete the option.
- `wc_remote_inbox_notifications_stored_state` has `new_product_count` set to 1, and according to the [line here](https://github.com/woocommerce/woocommerce-admin/blob/6acebc08cb7be11bce52801a94458ed302096c07/src/RemoteInboxNotifications/ProductCountRuleProcessor.php#L47), this was calculated to 4 instead of 3. I do have old products trashed, but not permanently deleted.
- When adding the third product, it seems that `as_enqueue_async_action` did not result in running the specs. I confirm that this [line runs](https://github.com/woocommerce/woocommerce-admin/blob/6acebc08cb7be11bce52801a94458ed302096c07/src/RemoteInboxNotifications/StoredStateSetupForProducts.php#L133). I've tried to wait for more than 5 minutes, navigated to other pages in wp-admin, but they did not trigger a spec run.

I'm not sure if I'm doing anything wrong, maybe if you could enlighten me @becdetat

becdetat committed:
Use init hook to initialize async AS action

becdetat committed:
Drop product count query workaround

becdetat commented:
Thanks @ilyasfoo 

I was hooking up the action-scheduler handler in the `admin_init` hook, which doesn't get executed on a WP Cron task, so the handler never got called most (perhaps 95%) of the time. It happened to work (when it didn't execute using WP Cron) when I was testing, so I thought it was ok before pushing this PR.

I've changed where the handler is hooked up (in the `init` hook with a priority of 0) and removed the product count workaround as you noticed (not required now since the product count processor happens in a different request to where the product is actually added).

ilyasfoo commented on src/RemoteInboxNotifications/ProductCountRuleProcessor.php:
```diff
@@ -40,14 +40,10 @@ public function __construct( $product_query = null ) {
-			? $stored_state->new_product_count
-			: 0;
-		$count             = $products->total + $new_product_count;
+		$products = $this->product_query->get_products();
```
Nice, thanks for making this fix! I thought it was a bit confusing. I was going to say to add a test for this change, but then I saw the existing tests and I think it's a simple enough logic to not require more tests.

ilyasfoo commented on src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php:
```diff
@@ -73,12 +76,21 @@ public static function on_admin_init() {
 
 		add_action( 'activated_plugin', array( __CLASS__, 'run' ) );
```
I wonder if it's worth considering making the engine run here async as well?

If this is agreed, I would also suggest adding the `woocommerce_activated_plugin` action to cater for [wcadmin embedded](https://github.com/woocommerce/woocommerce/pull/28698).

becdetat replied:
This is probably a good idea, I can't see a reason not to make it async apart from introducing complexity when there isn't a huge amount of gain - plugins would generally be activated much less than products are added/edited, and there isn't much much overhead in running the engine as it doesn't actually pull down from the feeds except on a first run when WCA is installed.

The extra action makes sense too. I'll create these as a follow-up ticket though.

ilyasfoo replied:
I agree, I think the added complexity might not worth the performance add. Thanks for making the follow-up ticket for the added action!

ilyasfoo reviewed (commented):
@becdetat I've tested this twice in a row and confirmed the async action works reliably both times. This is shaping up nicely! 

I have a suggestion to add async to the initial run when plugin activation as well, would love to know your thoughts, as I couldn't think of any case against it right now.

ilyasfoo reviewed (approved):
Really great work here, thanks! 🚢
